### ISSUE TITLE
Send post message to page

### ISSUE THREAD

**Issue by woodsbox** [chrome-php/headless-chromium-php#12]
Would it be possible to send a post message to a page?

**Comment by gsouf** [comment]
@woodsbox Do you mean use POST method with some body when calling ``$page->navigate``?

**Comment by woodsbox** [comment]
Well, I did a navigate to a page to fetch some data. And then I have to send a post message to a different page to receive the csv data. The cookies that are created by the navigate should also be send. So I think yes post method with some body when calling $page->navigate

**Comment by gsouf** [comment]
Ok usually sending post is only possible via a form submission or via ajax. 

Though I cannot guarantee it, I'm pretty sure this is doable when opening a new page. Give me some time to check.

**Comment by woodsbox** [comment]
do you think this can work?

document.querySelector("a[ng-click=""download()""]").click();

**Comment by gsouf** [comment]
Yes, you it should work, why dont you test by yourself in your real chrome browser? You could open the console a try this code directly on the page.

**Comment by woodsbox** [comment]
Thnx for tip I will do it like that.
Only question the form submit: $evaluation->waitForPageReload();
When I then do a: $content = $page->evaluate('document.documentElement.innerHTML')->getReturnValue();

Not all the javascript is loaded yet, any tips how to wait before that all is done?

Sorry if I'm spamming you to much!

**Comment by gsouf** [comment]
I guess that the reason it's that your javascript starts evaluating after the page has loaded. The simplest way to solve I think is to have a waiter method that would watch and wait for a given element to be populated. Can you raise a new issue for that please?

**Comment by woodsbox** [comment]
hi @gsouf  sorry for spamming you.
I have a small problem:
$page->navigate('somepage')->waitForNavigation();
$content = $page->evaluate('document.documentElement.innerHTML')->getReturnValue();

returns content where not all javascript is executed!
Any tip on how to wait?

**Comment by Kozin-br** [comment]
you might to use interceptions for making POST request

**Comment by ethaniel** [comment]
> you might to use interceptions for making POST request

This is exactly how they do it in Pupeteer: https://stackoverflow.com/questions/47060534/how-do-you-send-a-post-request-in-puppeteer
I'm trying to figure out a way how to port this to us.

**Comment by urinaldeuce** [comment]
@ethaniel Did you find a way to port this?
I have been attempting to use Guzzle for the intercept but this isn't my forte.

**Comment by ethaniel** [comment]
> @ethaniel Did you find a way to port this? I have been attempting to use Guzzle for the intercept but this isn't my forte.

Unfortunately, not. I ended up submitting the forms that were already presented to me on the page.
My guess is that it might be possible to create an HTML page with the needed form, make chromium open and submit it.

If chromium will pop up a confirm dialog, confirming the submission, it could be possibly handled with this code:
```php
$this->page->getSession()->on('method:Page.javascriptDialogOpening', function (array $params): void {
	echo "Dialog opened: " . $params["message"] . PHP_EOL;
	$this->page->getSession()->sendMessageSync(new Message('Page.handleJavaScriptDialog', ['accept' => true]));

});
```
